src/bsoncxx/builder/list.hpp

```diff
@@ -29,148 +30,119 @@ namespace builder {
+/// A JSON-like builder for creating BSON documents.
 ///
-class list {
-    using initializer_list_t = std::initializer_list<list>;
+class document {
+    using key_type = stdx::string_view;
+    using mapped_type = document;
+    using value_type = std::pair<key_type, mapped_type>;
+    using initializer_list_t = std::initializer_list<value_type>;
 
    public:
     ///
     /// Creates an empty document.
     ///
-    list() : list({}) {}
+    document() : document({}){};
 
     ///
-    /// Creates a bsoncxx::builder::list from a value of type T. T must be a
-    /// bsoncxx::types::bson_value::value or implicitly convertible to a
-    /// bsoncxx::types::bson_value::value.
+    /// Creates a bsoncxx::builder::document from a value of type T.
     ///
-    /// @param value
-    ///     the BSON value
+    /// @warning T must be a BSON type, i.e., implicitly convertible to a
+    /// bsoncxx::types::bson_value::value.
     ///
     /// @see bsoncxx::types::bson_value::value.
+    //
+    /// @warning Extracting a BSON document from an object created with this constructor will result
+    /// in an empty document. Use the key-value or std::initializer_list constructors instead.
+    ///
+    /// @note This constructor bridges the gap between a homogeneous container, a
+    /// std::initializer_list, and a heterogeneous one, a BSON document. In other words, the
+    /// std::initializer_list constructor implicitly uses this to convert all elements into a common
+    /// type, a bsoncxx::builder::document.
     ///
     template <typename T>
-    list(T value) : val{value} {}
+    document(T v) : _value{v} {}
 
     ///
-    /// Creates a BSON document, if possible. Otherwise, it will create a BSON array. A document is
-    /// possible if:
-    //      1. The initializer list's size is even; this implies a list of
-    //         key-value pairs or an empty document if the size is zero.
-    //      2. Each 'key' is a string type. In a list of key-value pairs, the 'key' is every other
-    //         element starting at the 0th element.
-    //
-    /// @param init
-    ///     the initializer list used to construct the BSON document or array
+    /// Creates a BSON document from a single key-value pair, e.g., { "key" : 1 }
     ///
-    /// @note
-    ///     to enforce the creation of a BSON document or array use the bsoncxx::builder::document
-    ///     or bsoncxx::builder::array constructor, respectively.
+    /// @param key
+    ///     the key in the key-value pair.
+    /// @param value
+    ///     the value in the key-value pair. It must be a BSON type, i.e., implicitly convertible to
+    ///     a bsoncxx::types::bson_value::value, or another bsoncxx::builder::document.
     ///
-    /// @see bsoncxx::builder::document
-    /// @see bsoncxx::builder::array
+    /// @see bsoncxx::types::bson_value::value.
     ///
-    list(initializer_list_t init) : list(init, true, true) {}
+    document(key_type key, mapped_type&& value)
```

Comment:
It looks like this constructor can only take a document as the value right now, but from reading your comments, it seems like you would like it to take any BSON type.  I think for that to work this would need to be templated:

```
template <typename T>
document(key_type key, const T& bson_value);

template <typename T>
document(key_type key, T&& bson_value);
```

^ I added a `const&` version, so the user can choose to either `std::move` in their value or let us copy it instead.

Reply:
Also, is this a constructor that we want to offer users?  It seems not that much simpler than just using the all-purpose list constructor below to create a one key-pair document:

```
builder::list::document builder("fruit", "apple");
builder::list::document builder2( {"fruit", "apple"});
```

I'd rather just offer the second way, because this shortcut constructor doesn't look list-like, and so is a bit confusing to see in this class.

Reply:
Updated this to take a generic type as opposed to the `mapped_type`.